packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -48,18 +49,58 @@ function findAllHotswappableChanges(
   const hotswappableResources = new Array<HotswapOperation>();
   let foundNonHotswappableChange = false;
   stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const nonHotswappableResourceFound = isNonHotswappableResourceChange(change);
     const lambdaFunctionShortCircuitChange = isHotswappableLambdaFunctionChange(logicalId, change, assetParamsWithEnv);
-    if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+    const stepFunctionShortCircuitChange = isHotswappableStepFunctionChange(logicalId, change, assetParamsWithEnv);
+
+    const requiresFullDeployment = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (stepFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (nonHotswappableResourceFound === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT);
+
+    const irrelevant = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (stepFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (nonHotswappableResourceFound === ChangeHotswapImpact.IRRELEVANT);
+
+    if (requiresFullDeployment) {
```

Comment:
I don't like this code structure. I think we need to do something much different here.

For each change, we call each "detector" function we have (right now, that's 2: `isHotswappableLambdaFunctionChange`, `isHotswappableStepFunctionChange()` - your new `isNonHotswappableResourceChange()` is a little bit different). We gather the results in an array.

Then, we run an algorithm, that goes something like this:

- If there are any `HotswapOperation`s in the array, take them out, and add them to the list of `hotswappableResources`. The algorithm ends there.
- If there are no `HotswapOperation`s in the array, check if there are any `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT`:
  - If there are, this means the entire result is `REQUIRES_FULL_DEPLOYMENT`, and we set the `foundNonHotswappableChange` variable to `true`.
  - If there are no `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` in the array, it means it's all `ChangeHotswapImpact.IRRELEVANT`, and so we continue on (just like if `isHotswappableLambdaFunctionChange()` would return `ChangeHotswapImpact.IRRELEVANT` today)

I might be missing some edge cases in the algorithm there, so it might need some polishing, but I hope the idea is clear 🙂.